DESIGN.md

```diff
@@ -0,0 +1,123 @@
+# RPC
+This project implements an RPC server in Go using gRPC as a transport.
+
+# Design
+What follows is a design document outlining the technology choices, architecture, and implementation of the RPC client and server. 
+
+## Authentication and Authorization
+The RPC service includes mutual TLS authentication. The client and server will both have their own private keys as well as certificates signed by the same CA. The minimum TLS version will be set to TLS1.2. In order to protect against downgrade attacks in TLS1.2, an allowlist will be used for ciphers. Conveniently, the Go TLS package will return a list of all cipher suites it implements excluding suites with known security issues (tls.CipherSuites()). In order to restrict the client to using these only, PreferServerCipherSuites will be set to true in the server TLS configuration. If this were a production application, an endpoint would be included in the API to accept a certificate revocation list, allowing the administrator to 'ban' certain clients. 
```

Comment:
> The minimum TLS version will be set to TLS1.2. In order to protect against downgrade attacks in TLS1.2, an allowlist will be used for ciphers. 

1. I'm probably just splitting a few hairs here... The allow list of ciphers will prevent against protocol downgrade attacks or this is what the minimum TLS version setting is for and the cipher allow list is for something else?

2. I'm curious, why choose TLS 1.2 vs other options?

> In order to restrict the client to using these only, PreferServerCipherSuites will be set to true in the server TLS configuration

This seems to imply setting PreferServerSipherSuites = false will not use the allow list? Is that the case?

Reply:
1. Some downgrade attacks, for example: FREAK, attempt to downgrade to a weaker cipher. If we set the TLS server to only accept secure ciphers than this class of downgrade attack will not be effective.
2. The oldest TLS protocol that is acceptable is TLS 1.2 [(source)](https://datatracker.ietf.org/doc/rfc8996/). One example of a critical security issue that prevents me from allowing TLS 1.1 or TLS1.0 is that they only offer vulnerable hashing algorithms for handshake MACs (MD5 and SHA1).
3. In the Go documentation for the TLS Config object it appears unnecessary to set PreferServerCipherSuites:
>  // CipherSuites is a list of supported cipher suites for TLS versions up to
>  // TLS 1.2. If CipherSuites is nil, a default list of secure cipher suites
    // is used, with a preference order based on hardware performance. The
    // default cipher suites might change over Go versions. Note that TLS 1.3
    // ciphersuites are not configurable.
    CipherSuites []uint16
>  
>  // PreferServerCipherSuites controls whether the server selects the
    // client's most preferred ciphersuite, or the server's most preferred
    // ciphersuite. If true then the server's preference, as expressed in
    // the order of elements in CipherSuites, is used.
    PreferServerCipherSuites bool // Go 1.1

To me it appears worthwhile to set it to true, at least to skip iterating over unacceptable cipher proposals offered by the client

Reply:
Ok, I think it's just terminology, on my part I tend to think of downgrade in the protocol version sense. TLS 1.2 is an upgrade over TLS1.1, so a downgrade attack is one that downgrades the version. You're referring to cipher selection within a protocol version, that is susceptible in the same way where a man in the middle can force the selection of a vulnerable cipher within a version.